data/transactions/logic/eval.go

```diff
@@ -685,9 +792,30 @@ func boolToUint(x bool) uint64 {
 	return 0
 }
 
-// MaxStackDepth should move to consensus params
+// MaxStackDepth should not change unless gated by a teal version change / consensus upgrade.
 const MaxStackDepth = 1000
 
+func (cx *EvalContext) remainingBudget() int {
+	if cx.runModeFlags == runModeSignature {
+		return int(cx.Proto.LogicSigMaxCost) - cx.cost
+	}
+	if cx.PooledApplicationBudget != nil {
+		return *cx.PooledApplicationBudget
+	}
+	return cx.Proto.MaxAppProgramCost - cx.cost
+}
+
+func (cx *EvalContext) remainingInners() int {
+	if cx.Proto.EnableInnerTransactionPooling && cx.pooledAllowedInners != nil {
+		return *cx.pooledAllowedInners
+	}
+	// Before EnableInnerTransactionPooling, MaxInnerTransactions was the amount
+	// allowed in a single txn. No consensus version should enable inner app
+	// calls without turning on EnableInnerTransactionPoolin, else inner calls
+	// could keep branching with "width" MaxInnerTransactions
+	return cx.Proto.MaxInnerTransactions - len(cx.Txn.EvalDelta.InnerTxns)
```

Comment:
I mean `remainingInners` are always used as `len(cx.subtxns) > cx.remainingInners()` where `len` is supposed to return a non-negative number, returning a non-negative from `remainingInners` does not leave chances for integer overflow

Reply:
len returns an int, so the comparison is signed.  This seems safe. I'm reluctant to switch to a sub saturate form, because then remainingInners() has a surprising interface. Someone might think that could make an inner call, and then check if remainingInner() < 0.  That is a sort of reasonable expectation.  But if it returns 0 at the lowest, it would be wrong.